File: conformance/wasm/shapes.mjs
// Conformance consumer: shapes sample, Wasm (wasm32-unknown-unknown) target.
//
// Drives the generated ESM bindings (loadWeaveffiWasm) against the real producer
// compiled to wasm, exercising rich (algebraic) enums as value types: plain
// tagged objects ({ tag: 'Circle', radius: 2.5 }) serialized into value
// buffers at the boundary, with no wrapper class, factories, or free().
// Covers every variant shape (unit, f64, two f32s, string + u8) through
// describe (enum in, string out) and scale (enum in and out), the consumer-side
// rejection of an unknown variant tag, plus the expanded numerics ([u8] in,
// u64 out as BigInt). Mirrors conformance/c/shapes.c.
//
// Inputs come from the harness:
//   WV_WASM: path to the compiled shapes.wasm
//   WV_JS:   path to the generated weaveffi_wasm.js (ESM)

import fs from 'fs';

const WASM = process.env.WV_WASM;
const JS = process.env.WV_JS;
if (!WASM || !JS) {
  console.error('WV_WASM and WV_JS must be set');
  process.exit(2);
}

// Node has no file:// fetch; shim it so the generated loader can read the .wasm.
globalThis.fetch = async (url) => ({ arrayBuffer: async () => fs.readFileSync(url) });

const mod = await import(JS);
const api = await mod.loadWeaveffiWasm(WASM);

let failures = 0;
function expect(cond, msg) {
  if (!cond) {
    console.error('assertion failed: ' + msg);
    failures++;
  }
}
function approx(a, b, eps) {
  return typeof a === 'number' && Math.abs(a - b) < eps;
}

// Plain C-style enum still crosses by value as a frozen discriminant object.
expect(mod.Channel && mod.Channel.Green === 1, 'plain enum Channel exported');

// describe: rich enum in (a plain tagged object), string out; dispatches on
// the active variant.
const circle = { tag: 'Circle', radius: 2.5 };
expect(api.shapes.describe(circle) === 'circle(r=2.5)', 'describe(circle)');
expect(api.shapes.describe({ tag: 'Empty' }) === 'empty', 'describe(empty)');
expect(
  /rect/.test(api.shapes.describe({ tag: 'Rectangle', width: 3.0, height: 4.0 })),
  'describe(rectangle)'
);
expect(
  /hex/.test(api.shapes.describe({ tag: 'Labeled', label: 'hex', count: 6 })),
  'describe(labeled)'
);

// scale: rich enum in and out. The circle's f64 payload round-trips scaled.
const big = api.shapes.scale(circle, 4.0);
expect(big.tag === 'Circle', "scaled.tag === 'Circle'");
expect(approx(big.radius, 10.0, 1e-9), 'scaled.radius == 10.0');

// Rectangle: two f32 payloads survive the round trip (scale by 2).
const rect2 = api.shapes.scale({ tag: 'Rectangle', width: 3.0, height: 4.0 }, 2.0);
expect(rect2.tag === 'Rectangle', "rect2.tag === 'Rectangle'");
expect(approx(rect2.width, 6.0, 1e-6), 'rect2.width == 6.0');
expect(approx(rect2.height, 8.0, 1e-6), 'rect2.height == 8.0');

// Labeled: the string + u8 payload passes through scale unchanged.
const labeled2 = api.shapes.scale({ tag: 'Labeled', label: 'hex', count: 6 }, 9.0);
expect(labeled2.tag === 'Labeled', "labeled2.tag === 'Labeled'");
expect(labeled2.label === 'hex', "labeled2.label == 'hex'");
expect(labeled2.count === 6, 'labeled2.count == 6');

// Empty: a unit variant is tag-only in both directions.
const empty2 = api.shapes.scale({ tag: 'Empty' }, 3.0);
expect(empty2.tag === 'Empty', "empty2.tag === 'Empty'");

// An unknown variant tag is a consumer-side marshalling failure, rejected
// before anything crosses the boundary.
let tagErr = null;
try { api.shapes.describe({ tag: 'Pentagon' }); } catch (e) { tagErr = e; }
expect(tagErr instanceof mod.WeaveFFIError, 'unknown tag -> WeaveFFIError');
expect(tagErr && tagErr.code === -3, `unknown tag -> marshalling code -3 (got ${tagErr && tagErr.code})`);
expect(tagErr && tagErr.message.includes('Pentagon'), 'unknown tag error names the tag');
expect(api.shapes.describe({ tag: 'Empty' }) === 'empty', 'module usable after the rejected tag');

// Floating-point edge values inside the rich enum payloads survive the round
// trip: NaN, infinities, and -0 in the f64 and f32 slots.
const nanCircle = api.shapes.scale({ tag: 'Circle', radius: NaN }, 1.0);
expect(Number.isNaN(nanCircle.radius), 'NaN radius survives');
const negZero = api.shapes.scale({ tag: 'Circle', radius: -0 }, 1.0);
expect(Object.is(negZero.radius, -0), '-0 radius keeps its sign');
const inf = api.shapes.scale({ tag: 'Rectangle', width: Infinity, height: -Infinity }, 1.0);
expect(inf.width === Infinity && inf.height === -Infinity, 'infinities survive in f32 slots');

// numerics: [u8] in (canonicalized to bytes), u64 out (BigInt);
// lowerCamelCase wrapper name.
expect(api.shapes.sumBytes([250, 250, 250, 250]) === 1000n, 'sumBytes == 1000n');
expect(typeof api.shapes.sumBytes([1]) === 'bigint', 'sumBytes returns a BigInt');
expect(api.shapes.sumBytes(new Uint8Array([255, 255])) === 510n, 'sumBytes accepts a Uint8Array');
expect(api.shapes.sumBytes([]) === 0n, 'sumBytes of nothing is 0n');

if (failures === 0) {
  console.log('wasm/shapes: OK');
} else {
  console.error(`wasm/shapes: ${failures} failure(s)`);
  process.exit(1);
}

File: conformance/kotlin/shapes.kt
// Conformance consumer: shapes sample, Android/Kotlin (JNI) target.
//
// Exercises the rich (algebraic) enum surface as value types: `Shape` is a
// generated sealed class (`Empty` object, `Circle`/`Rectangle`/`Labeled` data
// classes) that crosses JNI as a value buffer packed on the Kotlin side, so
// there are no native handles and nothing to close. Drives every variant
// through `describe` (buffered rich enum in, string out) and `scale` (buffered
// in and out, decoded back into the sealed hierarchy), covering unit
// (`Empty`), f64 (`Circle`), two f32 (`Rectangle`), and string + u8
// (`Labeled`) payloads, plus the C-style `Channel` enum's plain discriminants
// and the expanded numerics (`sumBytes`: bytes in, u64 out). Mirrors the C
// consumer's assertions. Compiled in-module with the generated `WeaveFFI.kt`.
@file:JvmName("Main")

import com.weaveffi.Channel
import com.weaveffi.Shape
import com.weaveffi.WeaveFFI
import kotlin.math.abs
import kotlin.system.exitProcess

fun expect(cond: Boolean, msg: String) {
    if (!cond) {
        System.err.println("assertion failed: $msg")
        exitProcess(1)
    }
}

fun main() {
    // describe dispatches on the active variant: each call packs the sealed
    // value into a buffer, and the producer formats it.
    expect(WeaveFFI.describe(Shape.Empty) == "empty", "describe empty")

    val circle = Shape.Circle(2.5)
    val circleDesc = WeaveFFI.describe(circle)
    expect(circleDesc == "circle(r=2.5)", "describe circle (got $circleDesc)")

    val rect = Shape.Rectangle(3.0f, 4.0f)
    val rectDesc = WeaveFFI.describe(rect)
    expect(rectDesc == "rectangle(3x4)", "describe rectangle (got $rectDesc)")

    val labeled = Shape.Labeled("hex", 6.toByte())
    val labeledDesc = WeaveFFI.describe(labeled)
    expect(labeledDesc == "labeled(hex x6)", "describe labeled (got $labeledDesc)")

    // scale: rich enum in and out; the return buffer decodes back into the
    // sealed hierarchy with the payload fields intact.
    val big = WeaveFFI.scale(circle, 4.0)
    expect(big is Shape.Circle, "scaled circle variant (got $big)")
    val bigRadius = (big as Shape.Circle).radius
    expect(abs(bigRadius - 10.0) < 1e-9, "scaled radius (got $bigRadius)")

    val bigRect = WeaveFFI.scale(rect, 2.0)
    expect(bigRect is Shape.Rectangle, "scaled rectangle variant (got $bigRect)")
    val r = bigRect as Shape.Rectangle
    expect(abs(r.width - 6.0f) < 1e-6f, "scaled width (got ${r.width})")
    expect(abs(r.height - 8.0f) < 1e-6f, "scaled height (got ${r.height})")

    // The string + u8 payload round-trips unchanged (scale leaves Labeled
    // alone), and data-class equality compares the decoded fields.
    val sameLabeled = WeaveFFI.scale(labeled, 2.0)
    expect(sameLabeled == labeled, "labeled round-trips (got $sameLabeled)")

    // The unit variant round-trips to the shared singleton.
    expect(WeaveFFI.scale(Shape.Empty, 2.0) == Shape.Empty, "empty round-trips")

    // A C-style enum keeps its plain int discriminants.
    expect(Channel.Green.value == 1, "Channel.Green == 1")
    expect(Channel.fromValue(2) == Channel.Blue, "Channel.fromValue(2) is Blue")

    // numerics: bytes in, u64 out. 250 wraps to a signed -6 byte but the JNI
    // shim reinterprets it as uint8_t, so the producer sums 250 * 4 == 1000.
    val total = WeaveFFI.sumBytes(
        byteArrayOf(250.toByte(), 250.toByte(), 250.toByte(), 250.toByte())
    )
    expect(total == 1000L, "sumBytes (got $total)")

    println("kotlin/shapes: OK")
}
